### PR TITLE
Implement "basic" Mastic

### PR DESCRIPTION
**Pull request by cjpatton**
Stacked on #22.

Add a new module to `poc` that implements the Mastic VDAF. This VDAF is similar to Poplar1 except that instead of aggregating counters, Mastic can aggregate any data type that is compatible with FlpGeneric.

Mastic operates in one of two "modes", depending on how the VIDPF is configured. In "incremental mode", the FLP is checked when evaluating at the first level of the VIDPF tree; in non-incremental mode, the FLP is checked when evaluating the last level of the tree (all intermediate levels are disabled).

When verifying the FLP, the encoded measurement (i.e., `beta`) is computed by evaluating the VIDPF at both child nodes of the root and summing the output shares. By one-hot verifiability, this sum is equal to value claimed by the Client; by path verifiability, checking this value is sufficient each node on the `alpha` path is guaranteed to have the same value.

The code is incomplete but is sufficient to illustrate the basic features we want. Some things to do:

* Add support for circuits with joint randomness

* Specify `PublicShare` and `InputShare`

* Decide whether to bind the query randomness to the VIDPF mode

* Implement message serialization

### CODE REVIEW

File: poc/vdaf_mastic.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+            # Evaluate the VIDPF at each child of the root node.
+            #
+            # It is sufficient to check the proof over the sum of the outputs,
+            # since the one-hot verifiability property of VIDPF ensures that
+            # exactly one of the children is equal to the encoded `beta` (and
+            # the other is equal to `0`).
+            #
+            # It is sufficient to check the proof just once, since the path
+            # verifiability property of VIDPF ensures that the same `beta` is
+            # propagated along the entire `alpha` path.
```
Comment (cjpatton) [inline poc/vdaf_mastic.py:126]: @jimouris please confirm that these "one-hot" and "path" verifiability are properties claimed by VIDPF.

File: poc/vdaf_mastic.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+            #
+            # Implementation note: This invocation of the VIDPF is redundant We
+            # evaluate at least one (and likely bot) of these prefixes during
+            # the main invocation below.
+            #
+            # TODO(cjpatton) Evaluate at the root instead, as long as this
+            # doesn't add communication cost.
```
Comment (cjpatton) [inline poc/vdaf_mastic.py:133]: @jimouris, @hannahdaviscrypto what would you like to do here? In the Overleaf version of the protocol, we evaluate the VIDPF "at the root" (i.e.,at the empty string $\varepsilon$), but this doesn't seem to be possible with the current VIDPF API. (I think Hannah was working on some changes to VIDPF that would have allowed this, but it wasn't compiling and I'm not sure how far she got with it.)

Evaluating at both children and then summing (as we do here) has a bit of CPU overhead. However it also has some advantages:
1. It's simple to specify
2. It has zero communication overhead. (If evaluating at the root, would we need to send more information in the public share so that the Aggregators can reconstruct shares of the output?)
3. It involves at most one more call to `Vidpf.eval_next()` than is necessary. (If `0` is not a prefix of any string in `prefixes`, then it'll be a waste to evaluate it.)

File: poc/vdaf_mastic.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+from flp_generic import FlpGeneric
+import hashlib
+from typing import Optional, Union
+from vdaf import Vdaf, test_vdaf
+from vdaf_poplar1 import Poplar1
+from vidpf import Vidpf
+from xof import XofShake128
+
+
+USAGE_QUERY_RAND = 0
```
Comment (jimouris) [inline poc/vdaf_mastic.py:14]: Nit: Can you add a comment to describe what these three are? Also, maybe an `enum` would be cleaner?

File: poc/vdaf_mastic.py [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+        # Generate FLP shares.
+        prove_rand = cls.Xof.expand_into_vec(cls.Field,
+            prove_rand_seed,
+            cls.domain_separation_tag(USAGE_PROVE_RAND),
+            b'',
+            cls.Flp.PROVE_RAND_LEN,
+        )
+
+        proof = cls.Flp.prove(beta, prove_rand, [])
```
Comment (jimouris) [inline poc/vdaf_mastic.py:63]: What is the last argument [] here?

File: poc/vdaf_mastic.py [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+        input_shares = [
+            (init_seed[0], leader_proof_share),
+            (init_seed[1], helper_seed),
+        ]
+        return (public_share, input_shares)
+
+    @classmethod
+    def is_valid(cls, agg_param, previous_agg_params):
+        (level, _prefixes) = agg_param
+        if cls.Vidpf.INCRIMENTAL_MODE:
```
Comment (jimouris) [inline poc/vdaf_mastic.py:76]: Rename to `INCREMENTAL_MODE`.

File: poc/vdaf_mastic.py (hunk 1 above)
Comment (jimouris) [reply poc/vdaf_mastic.py:126]: > It is sufficient to check the proof over the sum of the outputs, since the one-hot verifiability property of VIDPF ensures that exactly one of the children is equal to the encoded `beta` (and the other is equal to `0`).

Yeah, that's correct. Just to clarify, this holds if the proofs from both servers are equal.

> It is sufficient to check the proof just once since the path verifiability property of VIDPF ensures that the same `beta` is propagated along the entire `alpha` path.

The path verifiability does not rely on the VIDPF proofs, we get this from the PLASMA parent-children checks. I'd rephrase it to:
```
Path Verifiability: This property ensures that the same `beta` is propagated along the entire `alpha` path. Additionally, for all other evaluations that are not `alpha` (or prefixes of it), we get zero.
```

I'd also add the **One-hot Verifiability:** and **Path Verifiability:** in front of each to make it clear that we rely on these properties.

File: poc/vdaf_mastic.py [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+    @classmethod
+    def prep_shares_to_prep(cls, agg_param, prep_shares):
+        if len(prep_shares) != 2:
+            raise ValueError('unexpected number of prep shares')
+
+        # Verify the VIDPF output.
+        (level_proof_0, verifier_share_0) = prep_shares[0]
+        (level_proof_1, verifier_share_1) = prep_shares[1]
+        if level_proof_0 != level_proof_1:
+            raise Exception('output vector is not one-hot')
```
Comment (jimouris) [inline poc/vdaf_mastic.py:173]: How about?
```python
raise Exception('Path Verifiability Error: The output vector is not one-hot.')
```

File: poc/vdaf_mastic.py [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+                raise ValueError('prep share with missing verifier share')
+
+            verifier = vec_add(verifier_share_0, verifier_share_1)
+            if not cls.Flp.decide(verifier):
+                raise Exception('programmed measurement is not in range')
+
+        return None
+
+    @classmethod
+    def prep_next(Prio3, prep_state, prep_msg):
```
Comment (jimouris) [inline poc/vdaf_mastic.py:187]: Prio3?

File: poc/vdaf_mastic.py [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+        leader_proof_share = vec_sub(proof,
+                                     cls.helper_proof_share(helper_seed))
+
+        input_shares = [
+            (init_seed[0], leader_proof_share),
+            (init_seed[1], helper_seed),
+        ]
+        return (public_share, input_shares)
+
+    @classmethod
```
Comment (jimouris) [inline poc/vdaf_mastic.py:73]: Who runs `is_valid()`? I'm a bit confused about what it checks.

File: poc/vdaf_mastic.py [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+            # doesn't add communication cost.
+            (out_share, _level0_proof) = cls.Vidpf.eval(agg_id,
+                                                        correction_words,
+                                                        init_seed,
+                                                        0,
+                                                        [0, 1],
+                                                        cs_proofs,
+                                                        cls.ROOT_PROOF,
+                                                        nonce)
+            meas_share = vec_add(out_share[0], out_share[1])
```
Comment (jimouris) [inline poc/vdaf_mastic.py:142]: `out_share[0]` and `out_share[1]` are the left and right children, right? Then we add them into `meas_share` and check the FLP, correct? But the full evaluation happens in lines 105-112? Shouldn't we have a check like:

```python
if level == 0:
# check FLP
else:
# do parent-children checks
```

**Review by jimouris — CHANGES_REQUESTED** [review]
Looks good, thanks for putting it together! I added some suggestions, some questions, and some nits.

File: poc/vdaf_mastic.py (hunk 4 above)
Comment (cjpatton) [reply poc/vdaf_mastic.py:63]: joint randomness (currently empty)

File: poc/vdaf_mastic.py (hunk 2 above)
Comment (cjpatton) [reply poc/vdaf_mastic.py:133]: can you send a PR with the necessary change to `vidpf`?

File: poc/vdaf_mastic.py (hunk 6 above)
Comment (cjpatton) [reply poc/vdaf_mastic.py:173]: This isn't Path Verifiability though, right? This is One-hot Verifiability.

File: poc/vdaf_mastic.py (hunk 8 above)
Comment (cjpatton) [reply poc/vdaf_mastic.py:73]: See https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-07#section-5.3

File: poc/vdaf_mastic.py (hunk 2 above)
Comment (cjpatton) [reply poc/vdaf_mastic.py:133]: Err actually what do you mean by "latter"? If you want to evaluate at $\varepsilon$ (I agree this is cleaner) then I believe we need to change the API. Or am I misunderstanding something?

File: poc/vdaf_mastic.py (hunk 9 above)
Comment (cjpatton) [reply poc/vdaf_mastic.py:142]: I claim that you always want to do VIDPF verification and only do FLP verification once.

File: poc/vdaf_mastic.py (hunk 1 above)
Comment (jimouris) [reply poc/vdaf_mastic.py:131]: Correct. We only need 1 FLP to check \epsilon (or left and right) as you did, and then we'll need the parent-children checks to check that this beta value is propagated correctly. Something like: https://github.com/jimouris/mastic/blob/main/src/collect.rs#L246

File: poc/vdaf_mastic.py (hunk 2 above)
Comment (jimouris) [reply poc/vdaf_mastic.py:138]: I meant that I prefer evaluating the 2 children. VIDPF won't need any changes either.

File: poc/vdaf_mastic.py (hunk 6 above)
Comment (jimouris) [reply poc/vdaf_mastic.py:178]: Correct, I meant *One-hot* verifiability.

File: poc/vdaf_mastic.py (hunk 9 above)
Comment (jimouris) [reply poc/vdaf_mastic.py:147]: I might misunderstood something. By VIDPF verification you mean check the proofs? Yes, we should do that at every level. Additionally, we'd need to do the parent-children checks at every level to make sure that the beta value (which we verified is valid in level 0) is propagated correctly down the tree.

Comment (jimouris) [reply poc/vdaf_mastic.py:147]: Exactly. From Overleaf:

> So far, with VIDPF, the servers can verify that there is at most one non-zero output at each level. However, VIDPFs alone are not sufficient to guarantee that the keys are well formed. The servers still need to verify that: a) the non-zero values of the client are
across a single path in the binary tree, and b) the value of the root node is correctly propagated down the VIDPF tree. We formalize these requirements as an additional property: Path Verifiability: blah blah blah

The idea is that we want $y^γ - (y^{γ||0} + y^{γ||1}) = 0$. Then the following should hold:
$y_0^γ - y_0^{γ||0} - y_0^{γ||1} = - y_1^γ + y_1^{γ||0} + y_1^{γ||1}$. So each server computes their part (server 0 computes the left and server 1 computes the right) and they check if it's correct.

Comment (cjpatton) [reply poc/vdaf_mastic.py:147]: Ohhhh that's so beautifully simple. I think you already mentioned this at some point but it didn't bleed through.

To implement this, I think we need to to take @hannahdaviscrypto's changes to `vidpf`. I'll work on this.

Comment (jimouris) [reply poc/vdaf_mastic.py:147]: Thanks :) It might make more sense to do it on the mastic-vdaf level rather than in VIDPF. This might help as well: https://github.com/jimouris/mastic/blob/main/src/collect.rs#L246

**Review by jimouris — APPROVED** [review]

File: poc/vdaf_mastic.py (hunk 9 above)
Comment (cjpatton) [reply poc/vdaf_mastic.py:148]: I ended up extending VIDPF (see top-comment for justification): https://github.com/jimouris/draft-mouris-cfrg-mastic/pull/22

**Comment by cjpatton** [comment]
Rebased (on #22) and squashed.

File: poc/vdaf_mastic.py (hunk 1 above)
Comment (cjpatton) [reply poc/vdaf_mastic.py]: Done.

File: poc/vdaf_mastic.py [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+        (correction_words, cs_proofs) = public_share
+
+        # Ensure that candidate prefixes are all unique and appear in
+        # lexicographic order.
+        for i in range(1, len(prefixes)):
+            if prefixes[i-1] >= prefixes[i]:
+                raise ValueError('out of order prefix')
+
+        # Evaluate the VIDPF.
+        (out_share, level_proof) = cls.Vidpf.eval(agg_id,
```
[the remark this answers is not in the document]
Comment (cjpatton) [reply poc/vdaf_mastic.py:126]: We could eliminate it, but we'd need to make changes to `Vidpf`. In particular, we need to be able to compute our share of `beta`. Perhaps we could have `Vidpf` return `prefix_tree_share`?

[the remark this answers is not in the document]
Comment (jimouris) [reply poc/vdaf_mastic.py:126]: What do you mean, why? Isn't that the `out_share`?

[the remark this answers is not in the document]
Comment (cjpatton) [reply poc/vdaf_mastic.py:126]: To check the FLP we have to compute a share of `beta`. The way we do that here is by evaluating the VIDPF at the children of the root and adding them up.

**Review by jimouris — APPROVED** [review]

**Comment by cjpatton** [comment]
Rebased.

File: poc/vdaf_mastic.py [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+        (level, _prefixes) = agg_param
+        if cls.Vidpf.INCREMENTAL_MODE:
+            # The first level evaluated must be `0` and all levels must be distinct.
+            #
+            # TODO(cjpatton) Consider relaxing this check to allow for
+            # "fast-start". Rather than start at the first level, we might want
+            # to start at a later level and more candidate prefixes. For
+            # example, instead of starting at level `0` with all `2`-bit
+            # prefixes, we might start at level `7` with all `8`-bit prefixes.
+            #
```
Comment (hannahedavis) [inline poc/vdaf_mastic.py:89]: I think it might make more sense to do something like a 'fast-forward' approach rather than a fast-start. Even if we were to start at level 7 in this way, we'd still have to evaluate the entire tree and the FLP at level 0. So we could think about changing the agg_param format to a tuple of agg_params; is_valid would evaluate the same properties on the union of the tuples as it currently does. That way you could just evaluate levels 0 and 7 at once and then move on.

File: poc/vdaf_mastic.py [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+    @classmethod
+    def prep_shares_to_prep(cls, agg_param, prep_shares):
+        if len(prep_shares) != 2:
+            raise ValueError('unexpected number of prep shares')
+
+        # Verify the VIDPF output.
+        (level_proof_0, verifier_share_0) = prep_shares[0]
+        (level_proof_1, verifier_share_1) = prep_shares[1]
+        if level_proof_0 != level_proof_1:
+            raise Exception('output vector is not one-hot')
```
[the remark this answers is not in the document]
Comment (cjpatton) [reply poc/vdaf_mastic.py:191]: Please send a PR to fix since this got merged already!

File: poc/vdaf_mastic.py (hunk 11 above)
Comment (cjpatton) [reply poc/vdaf_mastic.py:89]: Sounds reasonable. At this stage I don't want to overcomplicate the spec, so I suggest documenting this idea in an issue. You might also consider updating this comment in a PR.

File: poc/vdaf_mastic.py [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+            # during the main invocation below.
+            (out_share, _level0_proof) = cls.Vidpf.eval(agg_id,
+                                                        correction_words,
+                                                        init_seed,
+                                                        0,
+                                                        [0, 1],
+                                                        cs_proofs,
+                                                        cls.ROOT_PROOF,
+                                                        nonce)
+            meas_share = vec_add(out_share[0], out_share[1])
```
[the remark this answers is not in the document]
Comment (cjpatton) [reply poc/vdaf_mastic.py:160]: Sounds reasonable.

File: poc/vidpf.py [hunk 14]
```diff
@@ -116,42 +121,46 @@ def eval(cls, agg_id, correction_words, init_seed, level, prefixes, cs_proofs, p
...
+        # Compute the path verifier.
+        sha3 = hashlib.sha3_256()
+        for prefix in prefixes:
+            for current_level in range(level):
+                node = prefix >> (level - current_level)
+                y =  prefix_tree_share[(node,        current_level)  ][2]
+                y0 = prefix_tree_share[(node<<1,     current_level+1)][2]
+                y1 = prefix_tree_share[((node<<1)|1, current_level+1)][2]
+                sha3.update(cls.Field.encode_vec(vec_sub(y, vec_add(y0, y1))))
+        path_proof = sha3.digest()
```
[the remark this answers is not in the document]
Comment (cjpatton) [reply poc/vidpf.py]: Right. I attempted to address with an "implementation note" above. Basically you can cache the `prefix_tree_share` across calls to `eval()`. Does this ake sense?

[the remark this answers is not in the document]
Comment (cjpatton) [reply poc/vidpf.py]: https://github.com/jimouris/draft-mouris-cfrg-mastic/blob/main/poc/vidpf.py#L109
